File: app/src/main/java/com/example/rossinyamaral/bank/statementsScreen/StatementsInteractor.java
package com.example.rossinyamaral.bank.statementsScreen;

import android.util.Log;

import com.example.rossinyamaral.bank.BankApi;
import com.example.rossinyamaral.bank.BankApplication;
import com.example.rossinyamaral.bank.model.StatementModel;

import java.util.List;

interface StatementsInteractorInput {
    public void fetchStatementsData(StatementsRequest request);
}


public class StatementsInteractor implements StatementsInteractorInput {

    public static String TAG = StatementsInteractor.class.getSimpleName();
    public StatementsPresenterInput output;
    public StatementsWorkerInput aStatementsWorkerInput;

    public StatementsWorkerInput getStatementsWorkerInput() {
        if (aStatementsWorkerInput == null) return new StatementsWorker();
        return aStatementsWorkerInput;
    }

    public void setStatementsWorkerInput(StatementsWorkerInput aStatementsWorkerInput) {
        this.aStatementsWorkerInput = aStatementsWorkerInput;
    }

    @Override
    public void fetchStatementsData(StatementsRequest request) {
        aStatementsWorkerInput = getStatementsWorkerInput();

        BankApplication.getInstance().bankApi.getStatements(request.userId,
                new BankApi.ApiCallback<List<StatementModel>>() {
                    @Override
                    public void onSuccess(List<StatementModel> object) {

                        StatementsResponse StatementsResponse = new StatementsResponse();
                        StatementsResponse.statements = object;
                        // Call the workers

                        output.presentStatementsData(StatementsResponse);
                    }

                    @Override
                    public void onError(BankApi.ErrorResponse message) {

                    }
                });
    }
}

File: bank_app/app/src/main/java/dev/vitorpacheco/solutis/bankapp/statementsScreen/StatementsInteractor.kt
package dev.vitorpacheco.solutis.bankapp.statementsScreen

interface StatementsInteractorInput {
    fun fetchStatementsData(request: StatementsRequest)
}

class StatementsInteractor : StatementsInteractorInput {

    var output: StatementsPresenterInput? = null
    var worker = StatementsWorker()

    override fun fetchStatementsData(request: StatementsRequest) {
        request.userId?.let {
            worker.listStatements(request) { output?.presentStatementsData(it) }
        }
    }

    companion object {
        var TAG = StatementsInteractor::class.java.simpleName
    }

}